cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,192 @@
+use crate::commands::rebase::rebase_descendants;
+use crate::ui::Ui;
+
+/// Parallelize revisions by making them siblings
+///
+/// The set of target commits being parallelized must have a single head and
+/// root commit (not to be confused with the repo root), otherwise the command
+/// will fail. Each of the target commits is rebased onto the parents of the
+/// root commit. The children of the head commit are rebased onto the target
+/// commits.
+///
+/// Example usage:
+/// ```text
+/// jj log
+/// ◉   3
+/// ◉   2
+/// ◉   1
+/// ◉
+/// jj parallelize 1::2
+/// jj log
+/// ◉     3
+/// ├─╮
+/// │ ◉   2
+/// ◉ │   1
+/// ├─╯
+/// ◉
+/// ```
+#[derive(clap::Args, Clone, Debug)]
+#[command(verbatim_doc_comment, visible_aliases=&["para"])]
+pub(crate) struct ParallelizeArgs {
+    revisions: Vec<RevisionArg>,
+}
+
+#[instrument(skip_all)]
+pub(crate) fn cmd_parallelize(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &ParallelizeArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let mut tx = workspace_command.start_transaction();
+    // The target commits are the commits being parallelized.
+    let target_commits: Vec<CommitId> = tx
+        .base_workspace_helper()
+        .parse_union_revsets(&args.revisions)?
+        .evaluate()?
+        .iter()
+        // We want parents before children, so we need to reverse the order.
+        .reversed()
+        .collect();
+    if target_commits.len() < 2 {
+        return Ok(writeln!(ui.status(), "Nothing changed.")?);
+    }
+    let target_revset = RevsetExpression::commits(target_commits.clone());
+    let target_head = get_head(&target_revset, tx.repo())?;
+    let target_root = get_root(&target_revset, tx.repo())?;
+    let connected_length = RevsetExpression::commits(vec![target_head, target_root.clone()])
+        .connected()
+        .evaluate_programmatic(tx.repo())?
+        .iter()
+        .count();
+    if connected_length != target_commits.len() {
+        return Err(user_error(
+            "Cannot parallelize since the target revisions are not connected.",
+        ));
+    }
+
+    // Rebase the children of each target commit onto its new parents. A child
+    // which had a target commit as an ancestor before parallelize ran will have
+    // the target commit as a parent afterward.
+    for target_commit in target_commits.iter() {
+        // These parents are shared by all children of the target commit and
+        // include the target commit itself plus any of its ancestors which are
+        // being parallelized.
+        let common_parents: IndexSet<CommitId> = target_revset
+            .intersection(&RevsetExpression::commit(target_commit.clone()).ancestors())
+            .evaluate_programmatic(tx.repo())?
+            .iter()
+            // We want parents before children, so we need to reverse the order.
+            .reversed()
+            .collect();
```

Comment:
Since we know it's a linear range, this can be simplified now, right? Something like `let common_parents = target_commits[..i].iter().collect();` (where you can get `i` by using `target_commits.iter().enumerate()` above).

Reply:
Yeah good point. This works great, and I suppose it's more efficient even if it's a little less readable and depends on the target_commits being in a specific order that matches the indexes from `enumerate()`. I tweaked a few things to make this work nicely.